tokio/src/sync/batch_semaphore.rs

```diff
@@ -0,0 +1,744 @@
+pub(crate) struct Semaphore {
+    waiters: Mutex<Waitlist>,
+    /// The current number of available permits in the semaphore.
+    permits: AtomicUsize,
+
+    /// Permits in the process of being released.
+    ///
+    /// When releasing permits, if the lock on the semaphore's wait list is held
+    /// by another task, the task releasing permits adds them to this counter.
+    /// The task holding the lock will continue releasing permits until the
+    /// counter is drained, allowing the `release` operation to be wait free.
+    ///
+    /// The first bit of this counter indicates that the semaphore is closing.
+    /// Therefore, all values are shifted over one bit.
+    adding: AtomicUsize,
+}
+
+struct Waitlist {
+    queue: LinkedList<Waiter>,
+    closed: bool,
+}
+
+/// Error returned by `Permit::try_acquire`.
+#[derive(Debug)]
+pub(crate) enum TryAcquireError {
+    Closed,
+    NoPermits,
+}
+/// Error returned by `Permit::poll_acquire`.
+#[derive(Debug)]
+pub(crate) struct AcquireError(());
+
+#[derive(Debug)]
+pub(crate) struct Permit {
+    state: PermitState,
+}
+
+pub(crate) struct Acquire<'a> {
+    node: Waiter,
+    semaphore: &'a Semaphore,
+    permit: &'a mut Permit,
+    num_permits: u16,
+}
+
+/// Permit state
+#[derive(Debug, Copy, Clone)]
+enum PermitState {
+    /// Currently waiting for permits to be made available and assigned to the
+    /// waiter.
+    Waiting(u16),
+
+    /// The number of acquired permits
+    Acquired(u16),
+}
+
+/// An entry in the wait queue.
+struct Waiter {
+    /// The current state of the waiter. 
+    /// 
+    /// This is either the number of remaining permits required by
+    /// the waiter, or a flag indicating that the waiter is not yet queued.
+    state: AtomicUsize,
+
+    /// The waker to notify the task awaiting permits.
+    /// 
+    /// # Safety
+    /// 
+    /// This may only be accessed while the wait queue is locked.
+    /// XXX: it would be nice if we could enforce this...
+    waker: CausalCell<Option<Waker>>,
+
+    /// Intrusive linked-list pointers.
+    /// 
+    /// # Safety
+    /// 
+    /// This may only be accessed while the wait queue is locked.
+    pointers: linked_list::Pointers<Waiter>,
+
+    /// Should not be `Unpin`.
+    _p: PhantomPinned,
+}
+
+const CLOSED: usize = 1 << 17;
+
+impl Semaphore {
+    /// Creates a new semaphore with the initial number of permits
+    pub(crate) fn new(permits: usize) -> Self {
+        assert!(permits <= std::u16::MAX as usize);
+        Self {
+            permits: AtomicUsize::new(permits),
+            waiters: Mutex::new(Waitlist {
+                queue: LinkedList::new(),
+                closed: false,
+            }),
+            adding: AtomicUsize::new(0),
+        }
+    }
+
+    /// Returns the current number of available permits
+    pub(crate) fn available_permits(&self) -> usize {
+        self.permits.load(SeqCst) & std::u16::MAX as usize
```

Comment:
This `SeqCst` is not doing anything as there are no `SeqCst` stores anywhere afaik.